Diagnostics/AVTester/Test-ExchAVExclusions.ps1

```diff
@@ -62,6 +78,86 @@ param (
 
+function CheckIfISAcceptedRootCA {
+    [CmdletBinding()]
+    param(
+        [string]$CAString,
+        [switch]$isFIPFS,
+        [switch]$Offline
+    )
+
+    switch ($CAString) {
+        'CN=Microsoft Corporate Root CA, O=Microsoft Corporation' { return $true }
+        'CN=Microsoft Root Authority, OU=Microsoft Corporation, OU=Copyright (c) 1997 Microsoft Corp.' { return $true }
+        'CN=Microsoft Root Certificate Authority, DC=microsoft, DC=com' { return $true }
+        'CN=Microsoft Root Certificate Authority 2010, O=Microsoft Corporation, L=Redmond, S=Washington, C=US' { return $true }
+        'CN=Microsoft Root Certificate Authority 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US' { return $true }
+        default {
+            if ($Offline -and $CAString -eq 'CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US' )
+            { return $true }
+            if ($isFIPFS) {
+                if ( $CAString -eq 'CN=VeriSign Class 3 Public Primary Certification Authority - G5, OU="(c) 2006 VeriSign, Inc. - For authorized use only", OU=VeriSign Trust Network, O="VeriSign, Inc.", C=US' )
+                { return $true }
+                if ( $Offline -and $CAString -eq 'CN="Oracle America, Inc.", OU=Code Signing Bureau, O="Oracle America, Inc.", L=Redwood Shores, S=California, C=US' )
+                { return $true }
+            }
+            return $false
+        }
+    }
+}
+
+function Test-UnknownCompany {
+    param (
+        [Parameter()]
+        [string]
+        $CompanyName
+    )
+
+    switch ($CompanyName) {
+        'Microsoft Corporation' { return $false }
+        'Microsoft Corporation.' { return $false }
+        'Microsoft' { return $false }
+        'Microsoft Corp.' { return $false }
+        'Microsoft CoreXT' { return $false }
+        'Корпорация Майкрософт' { return $false }
+        default { return $true }
+    }
+}
+
+function Test-UnknownModule {
+    param (
+        [Parameter()]
+        [string]
+        $ModuleName
+    )
+
+    switch ($ModuleName) {
+        'Microsoft.RightsManagementServices.Core.ni.dll' { return $false }
+        'Google.Protobuf.ni.dll' { return $false }
+        'Newtonsoft.Json.ni.dll' { return $false }
+        'l3codecp.acm' { return $false }
+        'ManagedBlingSigned.ni.dll' { return $false }
+        'System.IdentityModel.Tokens.jwt.ni.dll' { return $false }
+        default { return $true }
+    }
+}
+
+if ( (-not $DirectoriesAnalysis) -and (-not $ProcessesAnalysis) -and (-not  $ExtensionsAnalysis) ) {
+    $DirectoriesAnalysis = $true
+    $ProcessesAnalysis = $true
+    $ExtensionsAnalysis = $true
+}
+
+if ( $IncludeW3wpProcesses -and ( -not $ProcessesAnalysis ) ) {
+    Write-Error "IncludeW3wpProcesses parameter works with Processes Analysis"
+    exit
+}
+
+if ( $Recurse -and ( -not $DirectoriesAnalysis ) ) {
+    Write-Error "Recurse parameter works with Directories Analysis"
+    exit
+}
```

Comment:
This should be handled with `ParameterSetNames`

Reply:
Hi David,

I was thinking about that when I wrote the script, but I thought this solution is simpler.

If I choose the parameterSetNames I need to add many options and add complexity. 
That was what I tried:

    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryProcessesAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryExtensionsAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryProcessesExtensionsAnalysis')]
    [switch]
    $DirectoriesAnalysis,

    [Parameter(Mandatory = $false, ParameterSetName = 'NoSelectedAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryProcessesAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryProcessesExtensionsAnalysis')]
    [switch]
    $Recurse,

    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryProcessesAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryProcessesExtensionsAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'ProcessesAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'ProcessesExtensionsAnalysis')]
    [switch]
    $ProcessesAnalysis,

    [Parameter(Mandatory = $false, ParameterSetName = 'NoSelectedAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryProcessesExtensionsAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'ProcessesAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'ProcessesExtensionsAnalysis')]
    [switch]
    $IncludeW3wpProcesses,

    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryExtensionsAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'DirectoryProcessesExtensionsAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'ProcessesExtensionsAnalysis')]
    [Parameter(Mandatory = $true, ParameterSetName = 'ExtensionsAnalysis')]
    [switch]
    $ExtensionsAnalysis,

    [Parameter(Mandatory = $false, ParameterSetName = 'NoSelectedAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryProcessesAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryExtensionsAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'DirectoryProcessesExtensionsAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'ProcessesAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'ProcessesExtensionsAnalysis')]
    [Parameter(Mandatory = $false, ParameterSetName = 'ExtensionsAnalysis')]
    [switch]
    $OpenLog


And the message in some case of a wrong parameter is less clear, for example:
![image](https://github.com/microsoft/CSS-Exchange/assets/103440830/8b0e2c91-a3d1-48a3-be01-85780c7737b6)

But with the current script:
![image](https://github.com/microsoft/CSS-Exchange/assets/103440830/b2f9e620-008c-4bc0-b185-ae66b6393a15)

That was the reason why I decided not to use ParameterSetName. I do not know if it is possible with ParameterSetName in a different way.

What do you think?

Reply:
After looking at this a little more, I would almost go the reverse. Enable switches to disable things to check to make it more like `.\Test-ExchAVExclusions.ps1` without anything just does everything by default behavior. Which is what you are going for by doing this section of code:

![image](https://github.com/microsoft/CSS-Exchange/assets/22776718/b29d0595-17ae-4508-a7e9-c02a86a32862)

That would make it a little less complicated for end users to run. Then it wouldn't matter if `-Recurse` or `-IncludeW3wProcess` is enabled with the wrong set of parameters, because they also included a switch that deliberately disabled checks related to it.